docs/markdown/Writing Plugins/tutorials/02-integrate-goal-with-the-repository.md

```diff
@@ -0,0 +1,667 @@
+from pants.engine.rules import collect_rules, goal_rule, rule
+from pants.engine.target import (
+    HydratedSources,
+    HydrateSourcesRequest,
+    SourcesField,
+    Targets,
+)
+from pants.option.option_types import BoolOption
+from project_version.target_types import ProjectVersionSourceField, ProjectVersionTarget
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class ProjectVersionFileView:
+    path: str
+    version: str
+
+
+@rule
+async def get_project_version_file_view(
+    target: ProjectVersionTarget,
+) -> ProjectVersionFileView:
+    sources = await Get(HydratedSources, HydrateSourcesRequest(target[SourcesField]))
+    digest_contents = await Get(DigestContents, Digest, sources.snapshot.digest)
+    file_content = digest_contents[0]
+    return ProjectVersionFileView(
+        path=file_content.path, version=file_content.content.decode("utf-8").strip()
+    )
+
+
+class ProjectVersionSubsystem(GoalSubsystem):
+    name = "project-version"
+    help = "Show representation of the project version from the `VERSION` file."
+
+    as_json = BoolOption(
+        default=False,
+        help="Show project version information as JSON.",
+    )
+    check_git = BoolOption(
+        default=False,
+        help="Check Git tag of the repository matches the project version.",
+    )
+
+
+class ProjectVersionGoal(Goal):
+    subsystem_cls = ProjectVersionSubsystem
+
+
+class InvalidProjectVersionString(ValueError):
+    pass
+
+
+class ProjectVersionGitTagMismatch(ValueError):
+    pass
+
+
+class GitTagVersion(str):
+    pass
+
+
+@goal_rule
+async def goal_show_project_version(
+    console: Console,
+    targets: Targets,
+    project_version_subsystem: ProjectVersionSubsystem,
+) -> ProjectVersionGoal:
+    targets = [tgt for tgt in targets if tgt.has_field(ProjectVersionSourceField)]
+    results = await MultiGet(
+        Get(ProjectVersionFileView, ProjectVersionTarget, target) for target in targets
+    )
+    if project_version_subsystem.check_git:
+        git_repo_version = await Get(GitTagVersion, str, "")
+
+    for result in results:
+        try:
+            _ = Version(result.version)
+        except InvalidVersion:
+            raise InvalidProjectVersionString(
+                f"Invalid version string '{result.version}' from '{result.path}'"
+            )
+        if project_version_subsystem.check_git:
+            if git_repo_version != result.version:
+                raise ProjectVersionGitTagMismatch(
+                    f"Project version string '{result.version}' from '{result.path}' "
+                    f"doesn't match latest Git tag '{git_repo_version}'"
+                )
+
+        if project_version_subsystem.as_json:
+            console.print_stdout(json.dumps(dataclasses.asdict(result)))
+        else:
+            console.print_stdout(str(result))
+
+    return ProjectVersionGoal(exit_code=0)
+
+
+@rule
+async def get_git_repo_version(description: str = "") -> GitTagVersion:
```

Comment:
I am wondering if it's possible to have a rule without inputs at all? I had to pass something (here a description) just to satisfy the build graph.

Reply:
Seen this answered over in slack, but cross-posting for posterity:

Starting with Pants 2.15.x, you may call `Get` without any inputs like this:

https://github.com/pantsbuild/pants/blob/babf5dbb52342ed04c2fffbc559b66551b6fdd7d/src/python/pants/core/util_rules/environments.py#L697

Reply:
The docstring for `Get` has recently been fixed to mention this new syntax:

https://github.com/pantsbuild/pants/blob/babf5dbb52342ed04c2fffbc559b66551b6fdd7d/src/python/pants/engine/internals/selectors.py#L131-L156

Reply:
Done!